Compiler performance improvements around Symbols and Types

retronym opened the pull request:
Seems to bring about 4% in aggregate.

Thematically cherry-picked from #5785

_UPDATE_ : benchmark results, with the latest iteration of this PR.

Before:
```
$ for sha in 2.12.3-bin-9acab45-SNAPSHOT 2.12.3-bin-9b28b01-SNAPSHOT ; do sbt "set scalaVersion in compilation := \"$sha\"" "hot -psource= -f3 -wi 10 -jvmArgs -Xmx1G"; sleep 10; done

[info] Benchmark                                   (extraArgs)  (source)    Mode  Cnt     Score   Error  Units
[info] HotScalacBenchmark.compile                                         sample  240  1299.221 ± 6.034  ms/op
[info] HotScalacBenchmark.compile:compile·p0.00                           sample       1270.874          ms/op
[info] HotScalacBenchmark.compile:compile·p0.50                           sample       1291.846          ms/op
[info] HotScalacBenchmark.compile:compile·p0.90                           sample       1319.109          ms/op
[info] HotScalacBenchmark.compile:compile·p0.95                           sample       1354.655          ms/op
[info] HotScalacBenchmark.compile:compile·p0.99                           sample       1433.592          ms/op
[info] HotScalacBenchmark.compile:compile·p0.999                          sample       1434.452          ms/op
[info] HotScalacBenchmark.compile:compile·p0.9999                         sample       1434.452          ms/op
[info] HotScalacBenchmark.compile:compile·p1.00                           sample       1434.452          ms/op
```

After:

```
[info] Benchmark                                   (extraArgs)  (source)    Mode  Cnt     Score   Error  Units
[info] HotScalacBenchmark.compile                                         sample  251  1251.348 ± 7.449  ms/op
[info] HotScalacBenchmark.compile:compile·p0.00                           sample       1214.251          ms/op
[info] HotScalacBenchmark.compile:compile·p0.50                           sample       1243.611          ms/op
[info] HotScalacBenchmark.compile:compile·p0.90                           sample       1277.166          ms/op
[info] HotScalacBenchmark.compile:compile·p0.95                           sample       1335.466          ms/op
[info] HotScalacBenchmark.compile:compile·p0.99                           sample       1418.933          ms/op
[info] HotScalacBenchmark.compile:compile·p0.999                          sample       1451.229          ms/op
[info] HotScalacBenchmark.compile:compile·p0.9999                         sample       1451.229          ms/op
[info] HotScalacBenchmark.compile:compile·p1.00                           sample       1451.229          ms/op
```

retronym committed:
Avoid BoxesRuntime.equals in Type.unique

Like using AnyRefMap, we can optimize this datastructure with the knowledge
that is only holds AnyRef-s, and we can bypass BoxesRuntime for equality/hashing.

retronym committed:
Optimize TypeRef#equals

Avoid use of BoxesRuntime equals in favour of direct use
of Object#equals.

retronym committed:
Avoid List.apply in favor of ::

retronym committed:
Optimize TypeRef#computeHashCode

Directly call hashCode on the arguments, rather than
indirecting through ##.

retronym committed:
Optimize boundSyms by avoiding needless call to typeOfThis

This method is call in symbol substitution, which is a hot
path during typechecking.

This commit observes that certain types of types don't define
any bound symbols, and immediately returns an empty set.
These types of types are:

    ThisType     this.type
    SuperType    A.super.B
    ConstantType 42
    SingleType   foo.type
    TypeBounds   >: L <: B

When a type reports an empty set of bound symbols, symbol substution
_does_ recurse into the compoments of the type (with `mapOver`), so
we still find bound symbols, for instances, an existential type in
one of the type bounds.

Before this commit, calling `ThisType#boundSyms` came to the same
result via the implementation inherited from `SimpleTypeProxy`, which
incurred an needless call to `ThisType#typeOfThis`.

retronym committed:
Make Symbol.allOverridenSymbols tail recursive

retronym committed:
Optimize type comparison around AnnotatedTypes

Avoids the megamorphic call to `Type#annotations` in favour
of an instanceof check.

retronym committed:
Optimize copyTypeRef to avoid hash consing in some cases

For a no-args, no-prefix type ref, we can bypass the call
to TypeRef.apply, and with it, the hash-consing of the type,
by asking the Symbol for its type constructor ref, which is
cached.

retronym committed:
Opt specializedSym, special case for WildCardType

We find our way to this method with wildcard types in the
search for implicit views. This commit short circuits the
computation in this case, avoiding an as-seen-from operation
on the low symbol's info

retronym committed:
Avoid repeating call to Symbol#typeParams in hot code

retronym committed:
Add a fast path to TreeGen to speedup synthesis of casts

Unfortunately, we've had to add a lot more casts into the code
to workaround corner cases in the Fields transform. This commit
avoids an inefficiency in synthesizing these trees, by avoid
a call to as-seen-from.

retronym committed:
Avoid calling member type in futile matchingSymbolInternal check

If we arrive at this method with NoSymbol as the candidate symbol,
avoid an as-seen-from on the low side of the comparison.

l0rinc commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -696,21 +696,23 @@ trait Types
      */
     def memberInfo(sym: Symbol): Type = {
-//      assert(sym ne NoSymbol, this)
+      require(sym ne NoSymbol, this)
```
this seems to be more than an optimization
@adriaanm: https://github.com/paplorinc/scala/commit/1f6f7f8aa94c622665a35343de8108ea66a787b7#diff-ecaa4198a04261cd0e00a708b695eafbL689

l0rinc commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -1106,13 +1109,15 @@ trait Types
-*/
+    override def boundSyms: Set[Symbol] = emptySymbolSet
+
+    /*
```
minor: you may want to delete commented code instead of formatting it

l0rinc commented on src/reflect/scala/reflect/internal/util/Collections.scala:
```diff
@@ -61,6 +61,19 @@ trait Collections {
+    (thiss eq that) || {
+      var these = thiss
+      var those = that
+      while (!these.isEmpty && !those.isEmpty && these.head.equals(those.head)) {
```
minor: could probably be `while (these.nonEmpty && those.nonEmpty`

retronym replied:
I've got a bit of an aversion to `nonEmpty` (a mixed in trait method) in hot code because in some cases it is harder for JIT to devirtualize and inline. Not sure whether that's the case here.

l0rinc commented on src/reflect/scala/reflect/internal/util/Collections.scala:
```diff
@@ -61,6 +61,19 @@ trait Collections {
 
+  final def sameElementsEquals(thiss: List[AnyRef], that: List[AnyRef]): Boolean = {
+    // Probably immutable, so check reference identity first (it's quick anyway)
+    (thiss eq that) || {
```
question: this inlining of `thiss.sameElements(that)` seems like a micro-optimization to my untrained eye :)

retronym replied:
The key difference is using `.equals` rather than `.==`. The latter is routed through `BoxesRuntime.equals` which needs to perform a few type tests on the operands before delegating to `equals` in order to paper over the differences between primitive and boxed numeric equality.

l0rinc commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -2135,14 +2135,17 @@ trait Types
-        finalizeHash(mix(h, args.hashCode()), 3)
-      else
-        finalizeHash(h, 2)
+      var i = 0
```
could be inited with `2` to avoid `2 + i` later

l0rinc commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -2135,14 +2135,17 @@ trait Types
-      else
-        finalizeHash(h, 2)
+      var i = 0
+      var elem = args
```
minor: rename to `elems` instead?

l0rinc commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -829,7 +829,7 @@ self =>
       val arguments = right match {
         case Parens(args) => mkNamed(args)
-        case _            => List(right)
+        case _            => right :: Nil
```
the first is more readable, is the second that much faster?

retronym replied:
Yeah, we're lacking an optimization on `List.apply` to avoid boxing an array to pass as varargs. We've got a [ticket open](https://github.com/scala/scala-dev/issues/341) for that improvement, but in the meantime I'm happy enough with the readability of `::` to use it in the few places where this makes a material difference.

l0rinc commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -1076,7 +1076,8 @@ abstract class ClassfileParser {
         mod setInfo completer
         mod.moduleClass setInfo loaders.moduleClassLoader
-        List(cls, mod.moduleClass) foreach (_.associatedFile = file)
+        cls.associatedFile = file
```
if you want to retain the previous structure, you could probably
```scala
(cls.associatedFile, mod.moduleClass.associatedFile) = (file, file)
```

l0rinc commented on src/reflect/scala/reflect/internal/tpe/GlbLubs.scala:
```diff
@@ -184,8 +184,8 @@ private[internal] trait GlbLubs {
   /** Eliminate from list of types all elements which are a supertype
     *  of some other element of the list. */
   private def elimSuper(ts: List[Type]): List[Type] = ts match {
```
I don't claim to understand it, but I have a question:
This seems to be an `O(n^2)` operation (i.e. for every element search all remaining). Would it be faster to cache the parents in a `Set` and check the set for inconsistencies for the previously encountered parents instead?
That could make it linear, I think.
(would probably be slower if the number of elements is small)

retronym replied:
`Type`-s aren't amenable to use in a `HashSet` because `=:=` isn't compatible with `hashCode`.  This code here is actually using subtyping, which makes it harder again to find ways to beat `O(N^2)`.

l0rinc commented on src/reflect/scala/reflect/internal/tpe/GlbLubs.scala:
```diff
@@ -195,8 +195,8 @@ private[internal] trait GlbLubs {
     def elimSub0(ts: List[Type]): List[Type] = ts match {
-      case List() => List()
-      case List(t) => List(t)
+      case List() => ts
```
could probably be:
```scala
case List() | List(_) => ts
```
instead

l0rinc commented on src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:
```diff
@@ -101,11 +101,16 @@ private[internal] trait TypeMaps {
-            mapOverArgs(args, sym.typeParams)
-          else
+          if (trackVariance && args.nonEmpty && !variance.isInvariant) {
+            val tparams = sym.typeParams
```
minor: to make it more obvious what happened to the `sym.typeParams.nonEmpty` part of the `if`, this could probably be simplified to
```scala
if (sym.typeParams.nonEmpty) mapOverArgs(args, sym.typeParams)
else args mapConserve this
```

l0rinc commented on src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:
```diff
@@ -722,9 +722,20 @@ private[internal] trait TypeMaps {
       )
 
+    private def fromContains(syms: Set[Symbol]): Boolean = {
+      if (!syms.isEmpty) {
```
minor: can probably be `syms.nonEmpty && {`

l0rinc commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -2361,15 +2367,16 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
+      @tailrec
+      def loop(xs: List[Symbol], result: List[Symbol]): List[Symbol] = xs match {
+        case Nil     => result
         case x :: xs =>
```
minor: could also be `y :: ys` to avoid name collision with parent match

lrytz commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -3604,7 +3604,10 @@ trait Types
 
-      TypeRef(pre, sym, args)
+      if ((args eq Nil) && (pre eq NoPrefix))
+        sym.tpeHK // opt lean on TypeSymbol#tyconCache, rather than interning a type ref.
```
Could this be pushed further down into `TypeRef.apply`?

retronym replied:
IIRC, we can't (easily) do that because `tpeHK` calls that `TypeRef.apply`, so we'd loop.

lrytz replied:
OK, I see.. We could clean up TypeRef construction one day.

lrytz commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -1085,7 +1085,8 @@ trait Types
     override def baseTypeSeqDepth: Depth = supertype.baseTypeSeqDepth
     override def baseClasses: List[Symbol] = supertype.baseClasses
-  }
+    override def boundSyms: Set[Symbol] = emptySymbolSet
```
since `mapOver` recurses into components, it seems there are some unnecessary calls to `resultType.boundSyms`:
```
MethodType
    override def boundSyms = resultType.boundSyms ++ params

PolyType
    override def boundSyms = immutable.Set[Symbol](typeParams ++ resultType.boundSyms: _*)

```
and an unnecessary override in `NullaryMethodType`

```
    override def boundSyms = resultType.boundSyms
```

retronym replied:
Yep, I agree. Things seem to work with your proposed changes. I'm pushed a commit that actually avoids calling `boundSyms` altogether, which avoids the overhead of creating sets just to check if one of the `from` symbols exists in there, I'm pretty sure it will be faster to iterate over, e.g, `MethodSymbol#typeParams` directly and compare search for element to `from`.

We can actually short-circuit that search in many cases by pre-computing the range of symbol ids in `from` and by noting whether it has any term symbols.

lrytz replied:
Cool! `boundSyms` already confused me not long ago (I don't remember the details).

lrytz commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -3788,7 +3788,7 @@ trait Types
   private var uniqueRunId = NoRunId
 
-  protected def unique[T <: Type](tp: T): T = {
+  protected def unique[T <: Type](tp: T): T =  {
```
` `

lrytz commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -2928,6 +2938,34 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
     // are a case accessor (you can also be a field.)
     override def isCaseAccessorMethod = isCaseAccessor
 
+    def typeAsMemberOf(pre: Type): Type = {
```
```
Found 2 binary incompatibilities (22 were filtered out)
=======================================================
 * abstract synthetic method
   scala$reflect$runtime$SynchronizedSymbols$SynchronizedMethodSymbol$$$outer()scala.reflect.runtime.SynchronizedSymbols
   in interface
   scala.reflect.runtime.SynchronizedSymbols#SynchronizedMethodSymbol is
   present only in current version
 * abstract synthetic method
   scala$reflect$runtime$SynchronizedSymbols$SynchronizedMethodSymbol$$super$typeAsMemberOf(scala.reflect.internal.Types#Type)scala.reflect.internal.Types#Type
   in interface
   scala.reflect.runtime.SynchronizedSymbols#SynchronizedMethodSymbol is
   present only in current version
```

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -23,9 +23,6 @@ trait Namers extends MethodSynthesis {
   import global._
```
i cannot find the commit referenced in the commit message (8dc4dbf0edde0f83b76c6a31d1f741689fa2e67b), i guess it's 1f6f7f8aa94c622665a35343de8108ea66a787b7

lrytz commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -2928,6 +2938,34 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
+      if (mtpeRunId == currentRunId) {
+        if ((mtpePre eq pre) && (mtpeInfo eq info)) return {
+          mtpeResult
+        }
```
could merge the two `if`s, also the return block is a bit funny

lrytz commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -696,21 +696,23 @@ trait Types
+        computeMemberType(sym)
+    }
+
+    def computeMemberType(sym: Symbol): Type = sym.tpeHK match { //@M don't prematurely instantiate higher-kinded types, they will be instantiated by transform, typedTypeApply, etc. when really necessary
```
the comment can probably be dropped, since adriaan removed it in 1f6f7f8aa94c622665a35343de8108ea66a787b7

lrytz reviewed (commented):
Looking good!

retronym committed:
Minor refactor of TypeRef.computeHashCode

retronym committed:
Minor refactor in elimSuper pattern match

lrytz commented on src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:
```diff
@@ -703,18 +703,30 @@ private[internal] trait TypeMaps {
-        createFromClonedSymbols(bs, restp)((ps1, tp1) => PolyType(ps1, renameBoundSyms(tp1)))
-      case ExistentialType(bs, restp) =>
+      case MethodType(ps, restp) if fromHasTermSymbol && fromContains(ps) =>
+        createFromClonedSymbols(ps, restp)((ps1, tp1) => copyMethodType(tp, ps1, tp1))
```
I don't understand why we don't need the recursive calls to `renameBoundSyms` anymore.

retronym replied:
The result of this method is recursed into with `mapOver`, so we'll get back to here with the result type.

adriaanm replied:
Would be good to elaborate on this in a comment

retronym replied:
> The result of this method is recursed into with mapOver, so we'll get back to here with the result type.

Actually, taking another look at this, I think there is a potential problem with existential types, as `newExistentialType` collapses nested existential types. If some construction of existentials snuck past that, and an `ExistentialType(qs1, ExistentialType(qs2, res))` appeared here, we'd fail to look at `qs2`.

```scala
  /** A creator for existential types which flattens nested existentials.
   */
  def newExistentialType(quantified: List[Symbol], underlying: Type): Type =
    if (quantified.isEmpty) underlying
    else underlying match {
      case ExistentialType(qs, restpe) => newExistentialType(quantified ::: qs, restpe)
      case _                           => ExistentialType(quantified, underlying)
    }
```

As per @adriaanm's offline suggestion, I'm going to split this change into a separate PR and address the review comments there.

retronym replied:
Actually, the old code was already assuming existentials has been flattened.

retronym commented:
My measurements suggest that the last two commits (removing `boundSyms`, and avoiding futile comparisons of eta-expanded type refs in `<:<` and `=:=`) offer a further 2% speedup. 🏎

lrytz commented:
Very nice!

adriaanm commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -2931,6 +2943,31 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
+      // a local class symbol. We can no longer assume that`mtpePre eq pre` is a sufficient condition
+      // to use the cached result here.
+      //
+      // Rather than throwing away the baby with the bathwater, lets at least try to keep the caching
```
👍

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -696,21 +698,23 @@ trait Types
+      case OverloadedType(_, alts) =>
+        OverloadedType(this, alts)
       case tp =>
-        // Correct caching is nearly impossible because `sym.tpeHK.asSeenFrom(pre, sym.owner)`
```
Is this comment worth keeping around?

som-snytt replied:
The first thing I'm going to type tomorrow is `// is this comment worth keeping?`  I hope I don't accidentally read it six months from now, because it would blow my brain i.e. mind.  The other one I should plant for my future self is `// what were you thinking?`

retronym replied:
```
// Ceci n'est pas un commentaire
```

som-snytt replied:
`// comment?`

adriaanm commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4239,38 +4259,44 @@ trait Types
+      val symHiInfo = symHi.info
+      if (symHiInfo == WildcardType) {
+        if (symHi.isTerm) (!symHi.isStable || symLo.isStable)     // sub-member must remain stable
+        else true
```
Are we guaranteed `symHi.isAbstractType` in this case? Can an alias type's `info` ever be a wildcard? (If so, the lower symbol's info must also be a wildcard)

retronym replied:
The common case that I'm trying to optimize is when `symHi` is the member of a prototype for `HasMember` (e.g `{ def extensionMethod: ?}`)`. I'll tweak this to restrict this new fast path to `symHiInfo == WildcardType && symHi.isTerm` and let types flow down the the logic below.

retronym replied:
I've pushed this change now.

adriaanm commented on src/reflect/scala/reflect/internal/tpe/TypeComparers.scala:
```diff
@@ -360,10 +364,22 @@ trait TypeComparers {
+            val r1 = tp1.typeSymbolDirect.isNonBottomSubClass(tp2.typeSymbolDirect) // OPT faster than comparing eta-expanded types
+            val r2 = isSub(tp1.normalize, tp2.normalize)
+            if (r1 != r2)
+              getClass
```
stray debug artifact?

adriaanm commented on src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:
```diff
@@ -698,18 +703,30 @@ private[internal] trait TypeMaps {
     if (isDeveloper) assert(sameLength(from, to), "Unsound substitution from "+ from +" to "+ to)
 
+    private[this] var fromHasTermSymbol = false
+    private[this] var fromMin = Int.MaxValue
```
Kind of like a mini-bloom filter?

retronym replied:
Funny, I was re-reading about bloom filters while making this change. I guess this is kindred (we can cheaply determine non-existence in a set without iterating the full set), but OTOH we there isn't any superimposition of information.

I did learn about an [Zatocoding](https://en.wikipedia.org/wiki/Superimposed_code) in my travels, sort of an analog bloom filter for librarians.

retronym committed:
Rework changes to specializedSym

Restrict the fast path to term symbols (which is the important case
for performance) to allow the existing code to handle type symbols.

retronym commented:
I believe I've addressed the review comments here.

l0rinc reviewed (approved):
Awesome job, please update the description with the latest stats

adriaanm commented on src/reflect/scala/reflect/internal/tpe/TypeComparers.scala:
```diff
@@ -180,7 +180,11 @@ trait TypeComparers {
-    def retry(lhs: Type, rhs: Type) = ((lhs ne tp1) || (rhs ne tp2)) && isSameType(lhs, rhs)
+    def retry(lhs: Type, rhs: Type) = ((lhs ne tp1) || (rhs ne tp2)) && {
+      def sameSyms(tp1: Type, tp2: Type) = tp1.typeSymbolDirect eq tp2.typeSymbolDirect
+      val skipRetry = lhs.isInstanceOf[PolyType] && rhs.isInstanceOf[PolyType] && sameSyms(tp1, lhs) && sameSyms(tp2, rhs)
```
Could you convince me that we don't need to take into account potential differences in type params / type application due to normalization? I'm not sure how, but I'm also not sure it's impossible that `normalizePlus` could take e.g., `[T, U](T, U)` to `[T, U](U, T)` for one of `rhs`/`lhs`, so that now `[T,U]Swap[U,T] =:= [T,U]Tuple2[T,U]`

Also, this feels easier to follow:
```
      val skipRetry =
        (lhs eq tp1) && (rhs eq tp2) ||
          (lhs.isInstanceOf[PolyType] && rhs.isInstanceOf[PolyType] && sameSyms(tp1, lhs) && sameSyms(tp2, rhs))

      !skipRetry && isSameType(lhs, rhs)
```

retronym replied:
I've reworked that commit a to be more in line with the subtyping change (computing `skip` before actually calling `normalize`). Given that it is pretty subtle, let's take our time to review it in #5885, rather than in this omnibus.

adriaanm reviewed (commented):
The subtype change LGTM, but have a question on the sameType one.

retronym commented:
@paplorinc I've added the benchmark results to the description.

paplorinc commented:
@retronym the `cnt` seems to be different for the two runs, is that expected? I don't know how to compare them this way.

retronym commented:
The benchmark runs for a fixed duration, so after the speed up then count of executions is higher

The results suggest a 3.5% speed up.

paplorinc commented:
Thanks, good job!

retronym committed:
Merge remote-tracking branch 'origin/2.12.x' into faster/symtab

 Conflicts:
	src/reflect/scala/reflect/internal/Types.scala

Conflict was trivial related to commented out code.

retronym commented:
@adriaanm I've resolved the merge conflict here.

retronym commented:
ping @adriaanm for a (final?) review round here.

som-snytt commented:
I don't know what I'm going to do with my 3.5%. Probably catch up on Supergirl.
